### PR TITLE
Support workspaceFolders

### PR DESCRIPTION
**Pull request by hrsh7th**
This PR aims to support `workspaceFolders` capability.

For example, Sometimes the gopls does not return the completion items when editing multiple project's files.

```
/GOPATH
  /src
    /github.com
      /hrsh7th
        /vim-candle
          /go
            /candle-server
              go.mod
              main.go ... ①
      /mattn
        /gof
          main.go ... ②
```

Probably this behavior related to this code https://github.com/golang/tools/blob/9778d966fa8150a8a0027e8169749bae53a2ee86/internal/lsp/workspace.go#L19 .

### CODE REVIEW

**Comment by prabirshrestha** [comment]
Could you explain how this works?

**Comment by hrsh7th** [comment]
Sorry for my delay.
I explained the purpose for this PR.

File: autoload/lsp.vim [hunk 1]
```diff
@@ -390,7 +390,7 @@ function! s:ensure_flush(buf, server_name, cb) abort
     call lsp#utils#step#start([
         \ {s->s:ensure_start(a:buf, a:server_name, s.callback)},
         \ {s->s:is_step_error(s) ? s:throw_step_error(s) : s:ensure_init(a:buf, a:server_name, s.callback)},
-        \ {s->s:is_step_error(s) ? s:throw_step_error(s) : s:ensure_conf(a:buf, a:server_name, s.callback)},
+        \ {s->s:is_step_error(s) ? s:throw_step_error(s) : s:ensure_workspace(a:buf, a:server_name, s.callback)},
         \ {s->s:is_step_error(s) ? s:throw_step_error(s) : s:ensure_open(a:buf, a:server_name, s.callback)},
```
Comment (hrsh7th) [inline autoload/lsp.vim:394]: Before:
- Ensure only `configuration`

After:
- Ensure configuration and folders.

So I changed this name.

File: autoload/lsp.vim [hunk 2]
```diff
@@ -976,6 +970,13 @@ function! lsp#send_request(server_name, request) abort
         \)
 endfunction
 
+function! lsp#send_notification(server_name, params) abort
+    call s:send_notification(a:server_name, {
+    \   'method': a:params['method'],
+    \   'params': a:params['params'],
+    \ })
+endfunction
```
Comment (hrsh7th) [inline autoload/lsp.vim:978]: I was expose `s:send_notification` as public.

File: autoload/lsp.vim [hunk 3]
```diff
@@ -1035,25 +1036,8 @@ function! lsp#get_buffer_first_error_line() abort
     return lsp#ui#vim#diagnostics#get_buffer_first_error_line()
 endfunction
 
-function! s:merge_dict(dict_old, dict_new) abort
```
Comment (hrsh7th) [inline autoload/lsp.vim:1038]: I moved this method to utils.

**Comment by hrsh7th** [comment]
Sorry for my delay.

It ready to review now.

I checked below cases.

- The `json-languageserver` with `workspace_config`.
- The gopls's go-to-definition for other projects.

**Comment by hrsh7th** [comment]
I've heard the A still does not work on this PR so I try to make working it before merge.

**Pull request closed without merging**